Document how to transfer ownership of server side applied fields (#22412)
kubernetes/website: content/en/docs/reference/using-api/api-concepts.md

collaborator's changes. Conflicts can be forced, in which case the value will be
overridden, and the ownership will be transferred.

> When fields are removed from a configuration, fields which have no other owners

Reviewer: Possible reword, something like:
If you remove a field from a configuration, the field manager verifies that this field has no other owners.
If the field is not found in another object, it is deleted from the live object or reset to a default value.

Author: Thanks, I went with:

"If you remove a field from a configuration an apply it, server side apply checks
if there are any other field managers that also own the field.  If the field is
not found to be owned by any other field managers, it deleted from the live
object or reset to a default value. The same rule applies to associative list or
map items."

are deleted from the live object, or reset to their default value if they have
one. The same rule applies to associative list or map items.

> Apply is meant both as a replacement for the original `kubectl apply` and as a

Reviewer: nit: Server Side Apply is meant ...

collaborator's changes. Conflicts can be forced, in which case the value will be
overridden, and the ownership will be transferred.

> If you remove a field from a configuration an apply it, server side apply checks

Reviewer 2: ```suggestion
If you remove a field from a configuration and apply the configuration, server side apply checks
```

if there are any other field managers that also own the field.  If the field is

> not found to be owned by any other field managers, it deleted from the live

Reviewer 2: ```suggestion
not owned by any other field managers, it is either deleted from the live
```

> object or reset to a default value. The same rule applies to associative list or

Reviewer 2: ```suggestion
object or reset to its default value, if it has one. The same rule applies to associative list or
```

map items.

Server side apply is meant both as a replacement for the original `kubectl

> apply` and as a simpler mechanism to write controllers.

Reviewer 2: ```suggestion
apply` and as a simpler mechanism for controllers to enact their changes.
```

Although I'm not sure if we should say this yet, as we don't have good library support for it yet...

change a field which is managed by someone else will result in a rejected
request (if not forced, see [Conflicts](#conflicts)).

When two or more appliers set a field to the same value, they share ownership of
it. Any subsequent attempt to change the value of the shared field, by any of
the appliers, results in a conflict. Shared field owners may give up ownership
of a field by removing it from their configuration.

> Field management is stored in a newly introduced `managedFields` field that is

Author: Yeah, I'll remove.

change a field which is managed by someone else will result in a rejected
request (if not forced, see [Conflicts](#conflicts)).

When two or more appliers set a field to the same value, they share ownership of

> it. Any subsequent attempt to change the value of the shared field, by any of

Reviewer 2: ```suggestion
that field. Any subsequent attempt to change the value of the shared field, by any of
```

It is strongly recommended for controllers to always "force" conflicts, since they
might not be able to resolve or act on these conflicts.

### Transferring Ownership

In addition to the concurrency controls provided by [conflict
resolution](#conflicts), Server Side Apply provides ways to perform coordinated
field ownership transfers:

From a user to a controller: First the user signals to the controller that it
should take ownership of a field, usually by setting some other field that the
controller observes. The controller then claims ownership of a field by applying

> a config containing the field, with force set to true. Once the user observes

Reviewer: nit: s/config/configuration/

that the controller owns the field, they can remove the field from their applied
config to complete the transfer. This ensures the field is not accidently
removed from the live object, or reset to a default value, at any point of time
during the transfer. This can be used, for example, to transfer ownership of the

> `replicas` field from a user to a controller while enabling Horizonal Pod

Reviewer: nit: Horizontal

Autoscaling on an object. Controllers should use apply, not update, to claim

> ownership of fields since only apply claims ownership of field even if the

Reviewer: nit: ... of a field even if ...

desired value is the same as the live value.

From a user to another user: Users can transfer ownership of a field between
each other by setting the field to the same value in both of their applied

> configs, causing them to share ownership of the field. Once the users share

Reviewer: This is interesting.

resolution](#conflicts), Server Side Apply provides ways to perform coordinated
field ownership transfers from users to controllers.

First, the user delegates ownership of the field to a temporary, uniquely named,
field manager. Next, the user signals to the controller that it should take
ownership of the field, usually by setting some other field (perhaps in some
other object) that the controller observes. Last, controller then claims ownership
of the field by updating the object or applying a configuration containing the
field.

Transferring fields to controllers this way ensures the field is not accidently
removed from the live object, or reset to a default value, at any point of time
during the transfer.

This approach can be used, for example, to safely transfer ownership of the
`replicas` field from a user to a controller while enabling Horizontal Pod
Autoscaling on an object.

Say a user has defined deployment with `replicas` set to 100:

{{< codenew file="application/ssa/nginx-deployment.yaml" >}}

And the user created the deployment using server side apply like so:

```shell
kubectl apply -f application/ssa/nginx-deployment.yaml --server-side

> ```

Reviewer 2: Turn on HPA at this point (move that step here). It will motivate the next thing.

> Before the Horizontal Pod Autoscaler takes over the field and before the user

Reviewer 2: Hm, this still isn't quite right, other changes are actually safe, it's removing the replicas field that is unsafe. How about:

Now, the user would like to remove spec.replicas from their manifest, so they don't accidentally fight with the HPA controller. However, there is a race: it might take some time before HPA feels the need to adjust spec.replicas, and if the user removes spec.replicas before HPA writes to the field and becomes its owner, then apiserver will set spec.replicas to 1, its default value. This is not what the user wants to happen, even temporarily.

There are two solutions:

1. (easy) Leave spec.replicas in the manifest; when HPA eventually writes to that field, the system gives the user a conflict over it. At that point, it is safe to remove from the manifest.
2. (more advanced) If, however, the user doesn't want to wait, for example because they want to keep the cluster legible to coworkers, then they can take the following step to make it safe to remove spec.replicas from their manifest:

Author: Great. I've added this.

might not be able to resolve or act on these conflicts.

### Transferring Ownership

In addition to the concurrency controls provided by [conflict
resolution](#conflicts), Server Side Apply provides ways to perform coordinated
field ownership transfers:

From a user to a controller: First the user signals to the controller that it
should take ownership of a field, usually by setting some other field that the
controller observes. The controller then claims ownership of a field by applying
a configuration containing the field, with force set to true. Once the user
observes that the controller owns the field, they can remove the field from
their applied configuration to complete the transfer. This ensures the field is
not accidently removed from the live object, or reset to a default value, at any
point of time during the transfer. This can be used, for example, to safely transfer
ownership of the `replicas` field from a user to a controller while enabling
Horizontal Pod Autoscaling on an object. Controllers should use apply, not
update, to claim ownership of fields since only apply is able to claim ownership

> of a field even if the desired value is the same as the live value.

Reviewer 2: Describe what to do if the controller doesn't actually make a write to take ownership of the field, for example if it is already at the correct value. (Basically, the user instructions below, but just make up a unique manager name and have it only own the one field; when the controller eventually makes a write, the made up manager will no longer own anything and will be removed, so there's no clean up necessary.)

Author: Thanks. I've incorporated this. The user->controller section now looks like:

```
First the user signals to the controller that it should take ownership of a
field, usually by setting some other field that the controller observes. The
controller then claims ownership of the field by applying a configuration
containing the field. This can be done in one of the following ways:

- If the controller wants to immediately set the value to the field, it can
apply with force set to true.
- If the controller does not want to change the value to the field immediately,
the controller should genereate a unique, temporary, field manager name, and
claim shared ownership of the field by applying a configuration with the field
set to its live value. Later, when the controller does want to write the value
of the field, it can then apply with force set to true using the field manager
name it typically uses, which will automatically delete the temporary field
manager without the need for any clean up.

Once the user observes that the controller owns the field, they can remove the
field from their applied configuration to complete the transfer. This ensures
the field is not accidently removed from the live object, or reset to a default
value, at any point of time during the transfer.

This approach can be used, for example, to safely transfer ownership of the
`replicas` field from a user to a controller while enabling Horizontal Pod
Autoscaling on an object. Controllers should use apply, not update, to claim
ownership of fields since only apply is able to claim ownership of a field even
if the desired value is the same as the live value.
```

It is strongly recommended for controllers to always "force" conflicts, since they
might not be able to resolve or act on these conflicts.

### Transferring Ownership

In addition to the concurrency controls provided by [conflict
resolution](#conflicts), Server Side Apply provides ways to perform coordinated
field ownership transfers from users to controllers:

First the user signals to the controller that it should take ownership of a

> field, usually by setting some other field that the controller observes. The

Reviewer 2: ```suggestion
field, usually by setting some other field (perhaps in some other object) that the controller observes. The
```

> controller then claims ownership of the field by applying a configuration

Reviewer 2: ```suggestion
controller then claims ownership of the field by updating the object or applying a configuration
```

field, usually by setting some other field that the controller observes. The
controller then claims ownership of the field by applying a configuration
containing the field. This can be done in one of the following ways:

- If the controller wants to immediately set the value to the field, it can
apply with force set to true.
- If the controller does not want to change the value to the field immediately,

> the controller should genereate a unique, temporary, field manager name, and

Reviewer 2: It's the user that needs to do this step, not the controller. We can't assume controllers will be up to date.

Author: True... fixing.

Author: fixed

> - If the controller will not set the value of the field immediately, the user

Reviewer 2: This is the ordinary case right now, since no controller specifically uses SSA yet, and therefore it cannot take ownership of a field until it actually writes the object in a way that changes the field. 

So maybe this should go first?

Author: Yeah, let's lead with that.

Author: Fixed. I've also qualified the other case a bit more to make it clear that the user must be able to wait to observe the controller take ownership of the field.

of the field to it. This transfer should be done by first setting the field to
the same value in the applied configrations of the user and temporary field
manager, and then by removing it from the user's applied configuration after
field ownership is shared. Later, when the controller does want to set the value
of the field, it can then apply with force set to true, which will automatically
delete the temporary field manager without the need for any clean up.

> 

Reviewer 2: I hesitate to ask for more work, but I feel like a concrete example of handing off a deployment's replica count to an HPA would make this easier to understand for people?

Author: no problem, I'll add it.

Author: Restructured this section a bit to try to make the example flow: https://deploy-preview-22412--kubernetes-io-vnext-staging.netlify.app/docs/reference/using-api/api-concepts/#transferring-ownership

I'm testing the example on an apiserver now to make sure it actually works :)

Author: Works after fixing the example deployment file. Added a new example directory for server side apply so I can have example yaml that is not shared with other parts of the docs.

It is strongly recommended for controllers to always "force" conflicts, since they
might not be able to resolve or act on these conflicts.

### Transferring Ownership

In addition to the concurrency controls provided by [conflict
resolution](#conflicts), Server Side Apply provides ways to perform coordinated
field ownership transfers:

**From a user to a controller**: 

First the user signals to the controller that it should take ownership of a
field, usually by setting some other field that the controller observes. The
controller then claims ownership of the field by applying a configuration
containing the field. This can be done in one of the following ways:

- If the controller is designed to immediately set the value of the field, the
controller should apply with force set to true. Once the user observes that the
controller has taken ownership of the field, they can remove the field from
their applied configuration to complete the transfer.
- If the controller will not set the value of the field immediately, the user
should genereate a unique, temporary, field manager name, and transfer ownership

> of the field to it. This transfer should be done by first setting the field to

Reviewer 2: I'd move the "user to user" section above the user to controller section, so that here you can add:

```suggestion
of the field to it, similarly to the "user to user" instructions above. This transfer should be done by first setting the field to
```

the same value in the applied configrations of the user and temporary field
manager, and then by removing it from the user's applied configuration after
field ownership is shared. Later, when the controller does want to set the value
of the field, it can then apply with force set to true, which will automatically
delete the temporary field manager without the need for any clean up.

> Transferring fields ton controllers this way ensures the field is not accidently

Reviewer 2: ```suggestion
Transferring fields to controllers this way ensures the field is not accidentally
```

Reviewer 2: not fixed, missing push?

removed from the live object, or reset to a default value, at any point of time
during the transfer.

This approach can be used, for example, to safely transfer ownership of the
`replicas` field from a user to a controller while enabling Horizontal Pod
Autoscaling on an object. Controllers should use apply, not update, to claim
ownership of fields since only apply is able to claim ownership of a field even
if the desired value is the same as the live value.

> **From a user to another user**: 

Reviewer 2: Maybe actually move this section? :)

Author: Sounds good. I've made this a distinct markdown section, so the main section now focuses just on user->controller transfer.

resolution](#conflicts), Server Side Apply provides ways to perform coordinated
field ownership transfers from users to controllers.

First, the user delegates ownership of the field to a temporary, uniquely named,
field manager. Next, the user signals to the controller that it should take
ownership of the field, usually by setting some other field (perhaps in some
other object) that the controller observes. Last, controller then claims ownership
of the field by updating the object or applying a configuration containing the
field.

Transferring fields to controllers this way ensures the field is not accidently
removed from the live object, or reset to a default value, at any point of time
during the transfer.

This approach can be used, for example, to safely transfer ownership of the
`replicas` field from a user to a controller while enabling Horizontal Pod
Autoscaling on an object.

Say a user has defined deployment with `replicas` set to 100:

{{< codenew file="application/ssa/nginx-deployment.yaml" >}}

And the user created the deployment using server side apply like so:

```shell
kubectl apply -f application/ssa/nginx-deployment.yaml --server-side
```

> Before the user enables Horizontal Pod Autoscaling, the user needs to transfer

Reviewer 2: It's not before enabling HPA. You need to do this when all of 1) you enabled HPA, 2) HPA controller has not yet taken over the field, and 3) you need to make another change.

the `replicas` field to a temporary field manager by defining a configuration
that owns just the replicas field:

{{< codenew file="application/ssa/nginx-deployment-replicas-only.yaml" >}}

The user applies this configuration using a unique field manager name:

```shell

> kubectl apply -f application/ssa/nginx-deployment-replicas-only.yaml --server-side --field-manager=unique-manager-1s03n45l4n --validate=false

Reviewer 2: maybe "temporary" instead of "unique-manager"?

```

Next, the user removes the `replicas` field from original configuration:

{{< codenew file="application/ssa/nginx-deployment-no-replicas.yaml" >}}

The user then applies this configuration:

```shell
kubectl apply -f application/ssa/nginx-deployment-no-replicas.yaml --server-side
```

> Now the user may safely enable Horizontal Pod Autoscaling, e.g.:

Reviewer 2: No, now it's safe to make other changes.

Reviewer 2: Actually, it was always safe to make unrelated changes. It's removing the replicas field that was the potentially unsafe thing.

Author: Right. I've rephrased things so it's more clear that the replica field removal is the touchy part.

```
kubectl autoscale deployment nginx-deployment --cpu-percent=50 --min=1 --max=10
```

When the controller eventually sets the `replicas` field to a new value, the
temporary field manager will no longer own any fields and will be automatically

> deleted. No clean up is required.

Reviewer 2: An alternative workflow would be:
1. Enable the HPA, but *don't* remove the replicas field in your manifest
2. Don't make alterations to your workflow
3. When you eventually get a conflict over the replicas field, then remove it from your manifest.

This is less work for the user but more cognitive overhead, since there could be a reasonably large delay between 1 and 3.

Reviewer 2: (It might be good to present this as an alternative, just to help people form the correct mental model of the situation.)

Author: Yeah, this is what I'd expect to have happen to users if they don't proactively manage a transfer of the field. I'd expect adhoc enablement of HPA to end up in this workflow. I'll add it to the bottom of the section.

It is strongly recommended for controllers to always "force" conflicts, since they
might not be able to resolve or act on these conflicts.

### Transferring Ownership

In addition to the concurrency controls provided by [conflict
resolution](#conflicts), Server Side Apply provides ways to perform coordinated
field ownership transfers from users to controllers.

> This is best explained by example. Let's look at how to safely transfer

Reviewer 3: nit: [avoid using we](https://kubernetes.io/docs/contribute/style/style-guide/#avoid-using-we)

It's OK to postpone style fixes like that one until after this documentation PR merges.

ownership of the `replicas` field from a user to a controller while enabling

> Horizontal Pod Autoscaling on an object.

Reviewer 3: nit:
```suggestion
automatic horizontal scaling for a Deployment, using the HorizontalPodAutoscaler
resource and its accompanying controller.
```

It's OK to postpone style fixes like that one until after this documentation PR merges.

Say a user has defined deployment with `replicas` set to the desired value:

{{< codenew file="application/ssa/nginx-deployment.yaml" >}}

And the user has created the deployment using server side apply like so:

```shell
kubectl apply -f application/ssa/nginx-deployment.yaml --server-side
```

Then later, HPA is enabled for the deployment, e.g.:

```
kubectl autoscale deployment nginx-deployment --cpu-percent=50 --min=1 --max=10
```

Now, the user would like to remove `replicas` from their configuration, so they
don't accidentally fight with the HPA controller. However, there is a race: it
might take some time before HPA feels the need to adjust `replicas`, and if
the user removes `replicas` before the HPA writes to the field and becomes
its owner, then apiserver will set `replicas` to 1, its default value. This

> is not what the user wants to happen, even temporarily.

Reviewer 4: I really like how this is phrased 👍

Reviewer 3: nit: [address the reader as “you”](https://kubernetes.io/docs/contribute/style/style-guide/#address-the-reader-as-you)

It's OK to postpone style fixes like that one until after this documentation PR merges.

> There are two solutions:

Reviewer 3: If anyone is doing cleanup after this has merged, consider whether [tabs](https://kubernetes.io/docs/contribute/style/hugo-shortcodes/#tabs) would help here.

- (easy) Leave `replicas` in the configuration; when HPA eventually writes to that
  field, the system gives the user a conflict over it. At that point, it is safe
  to remove from the configuration.

> - (more advanced) If, however, the user doesn't want to wait, for example
>   because they want to keep the cluster legible to coworkers, then they can take
>   the following step to make it safe to remove `replicas` from their
>   configuration: Then, the user can remove the `replicas` field from original
>   configuration:

Reviewer 3: ```suggestion
- (more advanced) If, however, the user doesn't want to wait, for example
  because they want to keep the cluster legible to coworkers, then they can take
  the following steps to make it safe to remove `replicas` from their
  configuration:
```

and I also recommend using a Markdown ordered list rather than natural-language numbering (“First,” etc)

First, the user defines a new configuration containing only the `replicas` field:

{{< codenew file="application/ssa/nginx-deployment-replicas-only.yaml" >}}

> The user applies that configuration usine the field manager name `handover-to-hpa`:

Reviewer 3: ```suggestion
The user applies that configuration using the field manager name `handover-to-hpa`:
```

```shell
kubectl apply -f application/ssa/nginx-deployment-replicas-only.yaml --server-side --field-manager=handover-to-hpa --validate=false
```

> It is fine if this results in a conflict with the HPA controller. The conflict

Reviewer 4: Maybe it's just me (language barrier), but "it is fine" doesn't really tell me what I need to do (in this case nothing as opposed to force conflict).

Reviewer 3: I agree - this is a really difficult subject to teach and better wording would help a little.

indicates the controller has already claimed the field, which is what we want.

At this point the user may remove the `replicas` field from their configuration.

{{< codenew file="application/ssa/nginx-deployment-no-replicas.yaml" >}}

Note that whenever the HPA controller sets the `replicas` field to a new value,
the temporary field manager will no longer own any fields and will be
automatically deleted. No clean up is required.

### Transferring Ownership Between Users

Users can transfer ownership of a field between each other by setting the field
to the same value in both of their applied configs, causing them to share
ownership of the field. Once the users share ownership of the field, one of them
can remove the field from their applied configuration to give up ownership and

> complete the transfer to the other user.

Reviewer 3: I think this wording is OK. After this merges I will try to rephrase it, with the aim of making this understandable specifically to someone who is very new to Kubernetes and is trying to work out what details they can put off learning until later.

its owner, then apiserver will set `replicas` to 1, its default value. This
is not what the user wants to happen, even temporarily.

There are two solutions:

- (easy) Leave `replicas` in the configuration; when HPA eventually writes to that
  field, the system gives the user a conflict over it. At that point, it is safe
  to remove from the configuration.

- (more advanced) If, however, the user doesn't want to wait, for example
  because they want to keep the cluster legible to coworkers, then they can take
  the following steps to make it safe to remove `replicas` from their
  configuration:

First, the user defines a new configuration containing only the `replicas` field:

{{< codenew file="application/ssa/nginx-deployment-replicas-only.yaml" >}}

The user applies that configuration using the field manager name `handover-to-hpa`:

```shell
kubectl apply -f application/ssa/nginx-deployment-replicas-only.yaml --server-side --field-manager=handover-to-hpa --validate=false
```

If the apply results in a conflict with the HPA controller, then do nothing. The
conflict just indicates the controller has claimed the field earlier in the

> process that it sometimes does.

Reviewer 2: than
